scipy/linalg/tests/test_basic.py

```diff
@@ -21,9 +23,9 @@
 from scipy._lib._testutils import check_free_memory
 from scipy.linalg.blas import HAS_ILP64
 
-REAL_DTYPES = [np.float32, np.float64, np.longdouble]
-COMPLEX_DTYPES = [np.complex64, np.complex128, np.clongdouble]
-DTYPES = REAL_DTYPES + COMPLEX_DTYPES
+REAL_DTYPES = Union[np.float32, np.float64, np.longdouble]
+COMPLEX_DTYPES = Union[np.complex64, np.complex128, np.clongdouble]
+DTYPES = Union[REAL_DTYPES, COMPLEX_DTYPES]
```

Comment:
```suggestion
REAL_DTYPES: List[type] = [np.float32, np.float64, np.longdouble]
COMPLEX_DTYPES: List[type] = [np.complex64, np.complex128, np.clongdouble]
DTYPES = REAL_DTYPES + COMPLEX_DTYPES
```

Reply:
So what exactly is the goal of these constants?
Are they meant to represent type aliases or a sequence of types?

In the former case the original the original code was correct, in the latter I'd suggest using (unannotated) tuples rather than lists (as tuples can easily be used for expressing inhomogeneous sequences).

Reply:
It's not a type annotation, it's a list of dtypes used in testing.
```
for dtype in REAL_DTYPES:
    # run test for dtype
```

Reply:
```suggestion
REAL_DTYPES = (np.float32, np.float64, np.longdouble)
COMPLEX_DTYPES = (np.complex64, np.complex128, np.clongdouble)
DTYPES = REAL_DTYPES + COMPLEX_DTYPES
```

Right, If so I'd suggest the changes below (non-explicitly annotated tuples rather than lists).

The advantage of using tuples here is that type checkers can link each `DTYPES` element to a particular `np.inexact` subclass, which is not possible with lists as they can only be annotated as homogeneous sequences.

``` python
from typing import TYPE_CHECKING
import numpy as np

REAL_DTYPES = (np.float32, np.float64, np.longdouble)
COMPLEX_DTYPES = (np.complex64, np.complex128, np.clongdouble)
DTYPES = REAL_DTYPES + COMPLEX_DTYPES

if TYPE_CHECKING:
    # Revealed type is 'numpy.floating[numpy.typing._32Bit*]'
    reveal_type(REAL_DTYPES[0]())

    # Revealed type is 'numpy.complexfloating[numpy.typing._32Bit*, numpy.typing._32Bit*]'
    reveal_type(COMPLEX_DTYPES[0]())

    # Revealed type is 'numpy.complexfloating[numpy.typing._64Bit*, numpy.typing._64Bit*]'
    reveal_type(DTYPES[4]())
```

Reply:
From the build : 
```
scipy/linalg/tests/test_basic.py:32: note: Revealed type is 'numpy.floating[numpy.typing._32Bit*]'
scipy/linalg/tests/test_basic.py:35: note: Revealed type is 'numpy.complexfloating[numpy.typing._32Bit*, numpy.typing._32Bit*]'
scipy/linalg/tests/test_basic.py:38: note: Revealed type is 'numpy.complexfloating[numpy.typing._64Bit*, numpy.typing._64Bit*]'
```

I think I am not putting the condition snippet in the code, am I?

Reply:
> I think I am not putting the condition snippet in the code, am I?

Correct, this was merely intended as an example.